cfpb/sbl-frontend
# [Update Institution Profile] Form validations

Phase 4 of #222 

## Changes
- Eliminates the majority of ESLint exception previously applied in the UpdateFinancialProfile folder
- Updates type structures used to manage the UFP form
- Validates TIN
- Validates RSSD ID
- Ensures at least one Type of financial institution is selected
- Ensures that if Type - `Other` is selected, a description is entered in the checkbox
- Clears the description textbox and disables it if Type - `Other` is deselected
- Ensures *some* Institution data has been modified before allowing the user to submit the form
  - Note: Disabling the `Submit` button requires more extensive modifications due to struggles I've had getting `react-hook-form` to recognize changes to the `Type of financial institution` and `Additional details` fields.

## Questions
- [x] Are current error messages sufficient? 
  - ~~Particularly for RSSD ID: `Expected number, received nan`~~
  - Updated to: `RSSD ID must be a number`
- [x] What validations are needed for `Top Holder` and `Parent Entity`
  - ex. If LEI/RSSDID are populated, does Name have to be populated? (not implemented)
  - Update: No additional validations for now
- [x] Can a user submit the form with only `Additional details` populated?
  - Currently we block that
  - Update: Will continue to block for now.  Need to determine if this `Update` form should be allowed for other purposes

## How to test this PR

1. Visit http://localhost:8899/institution/123456789TESTBANK123/update
2. Click `Submit` without making any changes, you'll see a warning
3. Change TIN to an invalid entry i.e. `88-912312` and hit `Submit`, see that you get an error
4. Delete content in RSSD ID and hit `Submit`, see that you get an error
5. Uncheck all `Type(s) of financial institution` boxes and hit `Submit`, see that you get an error



## Screenshots

### No changes made, submission prevented
<img width="810" alt="Screenshot 2024-03-22 at 2 27 02 PM" src="https://github.com/cfpb/sbl-frontend/assets/2592907/19a313c3-892e-46cf-942c-334c73c99062">

### Invalid TIN, RSSD ID, and no Type(s) selected
<img width="783" alt="Screenshot 2024-03-26 at 1 45 57 PM" src="https://github.com/cfpb/sbl-frontend/assets/2592907/fca725f1-7e19-489e-86aa-59e7f0e2179c">


### Type - Other selected but not described
<img width="794" alt="Screenshot 2024-03-22 at 2 28 48 PM" src="https://github.com/cfpb/sbl-frontend/assets/2592907/1be3767b-5491-4564-bc24-2f9ee72cdf6f">
<img width="786" alt="Screenshot 2024-03-22 at 2 28 56 PM" src="https://github.com/cfpb/sbl-frontend/assets/2592907/99616a0c-7e58-4c70-a997-ca10d46af3f8">
